wandb/sdk/lib/mailbox.py

```diff
@@ -0,0 +1,125 @@
+"""
+mailbox.
+"""
+
+import secrets
+import string
+import threading
+import time
+from typing import Callable, Dict, Optional
+
+from wandb.proto import wandb_internal_pb2 as pb
+
+
+def _generate_address(length: int = 12) -> str:
+    address = "".join(
+        secrets.choice(string.ascii_lowercase + string.digits) for i in range(length)
+    )
+    return address
+
+
+class _MailboxSlot:
+    _result: Optional[pb.Result]
+    _event: threading.Event
+    _lock: threading.Lock
+    _address: str
+
+    def __init__(self, address: str) -> None:
+        self._result = None
+        self._event = threading.Event()
+        self._lock = threading.Lock()
+        self._address = address
+
+    def _get_and_clear(self, timeout: float) -> Optional[pb.Result]:
+        found = None
+        with self._lock:
+            if self._event.wait(timeout=timeout):
+                found = self._result
+                self._event.clear()
+        return found
+
+    def _deliver(self, result: pb.Result) -> None:
+        with self._lock:
+            self._result = result
+            self._event.set()
+
+
+class MailboxProgressHandle:
+    _percent_done: float
+
+    def __init__(self, percent_done: float) -> None:
+        self._percent_done = percent_done
+
+    @property
+    def percent_done(self) -> float:
+        return self._percent_done
+
+
+class MailboxHandle:
+    _mailbox: "Mailbox"
+    _slot: _MailboxSlot
+
+    def __init__(self, mailbox: "Mailbox", slot: _MailboxSlot) -> None:
+        self._mailbox = mailbox
+        self._slot = slot
+
+    def wait(
+        self,
+        timeout: Optional[float] = None,
+        on_progress: Callable[[MailboxProgressHandle], None] = None,
+        release: bool = True,
+    ) -> Optional[pb.Result]:
+        found: Optional[pb.Result] = None
+        start_time = time.time()
+        percent_done = 0.0
+        while True:
```

Comment:
so there is a risk of infinite loop here?
if we don't get found back and don't set the timeout... there is no way to break out of the infinite loop...

Reply:
This is true. i was thinking that there are some cases where we would want to wait forever.. for example run.finish(). we have no enforced wait time from the caller as we are trying to make sure data is synced...
 
I think it might be a good idea to change the default though. or make certain operations not allowed... like maybe progress indicator is required if you dont have a timeout.   I will think more about these defaults and incorporate that into the phase2 PR.  right now i suspect it is either a timeout or progress callback needs to be specified.

Reply:
yeah that's exactly what i was thinking... that the default should be something that times out but i like the idea of either progress or timeout...